File: tools/gen_grapheme.py
#!/usr/bin/env python3
"""Generate src/unicode_grapheme_data.zig from the UCD grapheme-break data."""
import sys, re

def parse_ranges(path, want_prop=None, field=1):
    out = []
    for line in open(path):
        line = line.split('#')[0].strip()
        if not line:
            continue
        parts = [p.strip() for p in line.split(';')]
        if len(parts) <= field:
            continue
        prop = parts[field]
        if want_prop is not None and prop != want_prop:
            continue
        rng = parts[0]
        if '..' in rng:
            lo, hi = rng.split('..')
        else:
            lo = hi = rng
        out.append((int(lo, 16), int(hi, 16), prop))
    return out

# GraphemeBreakProperty: class per range.
CLASS = {'CR':'cr','LF':'lf','Control':'control','Extend':'extend','ZWJ':'zwj',
         'Regional_Indicator':'ri','Prepend':'prepend','SpacingMark':'spacingmark',
         'L':'l','V':'v','T':'t','LV':'lv','LVT':'lvt'}
gb = []
for lo, hi, prop in parse_ranges('/tmp/gbp.txt'):
    if prop in CLASS:
        gb.append((lo, hi, CLASS[prop]))
gb.sort()

# Extended_Pictographic from emoji-data.txt.
extpict = [(lo, hi) for lo, hi, _ in parse_ranges('/tmp/emoji.txt', 'Extended_Pictographic')]
extpict.sort()

# InCB (Indic Conjunct Break) = Linker | Consonant | Extend from DerivedCoreProperties.
incb = []
for line in open('/tmp/dcp.txt'):
    line = line.split('#')[0].strip()
    if not line or 'InCB' not in line:
        continue
    parts = [p.strip() for p in line.split(';')]
    # format: range ; InCB ; Linker
    if len(parts) < 3 or parts[1] != 'InCB':
        continue
    kind = {'Linker':'linker','Consonant':'consonant','Extend':'incb_extend'}.get(parts[2])
    if not kind:
        continue
    rng = parts[0]
    lo, hi = (rng.split('..') if '..' in rng else (rng, rng))
    incb.append((int(lo,16), int(hi,16), kind))
incb.sort()

def emit(name, rows, has_kind):
    L = [f"pub const {name} = [_]{'GB' if has_kind else 'Range'}{{"]
    for r in rows:
        if has_kind:
            L.append('    .{ .lo = 0x%X, .hi = 0x%X, .v = .%s },' % (r[0], r[1], r[2]))
        else:
            L.append('    .{ .lo = 0x%X, .hi = 0x%X },' % (r[0], r[1]))
    L.append("};")
    return "\n".join(L)

out = []
out.append("//! GENERATED from UCD GraphemeBreakProperty.txt, emoji-data.txt, and")
out.append("//! DerivedCoreProperties.txt (InCB) — tools/gen_grapheme.py. Sorted ranges")
out.append("//! for the UAX #29 extended grapheme cluster algorithm.")
out.append("")
out.append("pub const Class = enum { other, cr, lf, control, extend, zwj, ri, prepend, spacingmark, l, v, t, lv, lvt };")
out.append("pub const Incb = enum { linker, consonant, incb_extend };")
out.append("pub const GB = struct { lo: u21, hi: u21, v: Class };")
out.append("pub const Range = struct { lo: u21, hi: u21 };")
out.append("pub const IncbR = struct { lo: u21, hi: u21, v: Incb };")
out.append("")
out.append(emit("gb_class", gb, True))
out.append("")
out.append(emit("ext_pictographic", extpict, False))
out.append("")
# incb uses IncbR
L = ["pub const incb = [_]IncbR{"]
for r in incb:
    L.append('    .{ .lo = 0x%X, .hi = 0x%X, .v = .%s },' % (r[0], r[1], r[2]))
L.append("};")
out.append("\n".join(L))
out.append("")
sys.stdout.write("\n".join(out))
sys.stderr.write("gb=%d extpict=%d incb=%d\n" % (len(gb), len(extpict), len(incb)))

File: tools/gen_grapheme.ts
/** Generate UAX #29 grapheme-break tables from UCD inputs. */
import { readText } from "./lib/home";

type Range = [number, number, string];
function parseRanges(path: string, wanted?: string, field = 1): Range[] {
  const rows: Range[] = [];
  for (const raw of readText(path).split("\n")) {
    const line = raw.split("#")[0].trim();
    if (!line) continue;
    const parts = line.split(";").map(value => value.trim());
    if (parts.length <= field) continue;
    const property = parts[field];
    if (wanted != null && property !== wanted) continue;
    const bounds = parts[0].split("..");
    rows.push([Number.parseInt(bounds[0], 16), Number.parseInt(bounds.length > 1 ? bounds[1] : bounds[0], 16), property]);
  }
  return rows;
}
const classes: Record<string, string> = { CR: "cr", LF: "lf", Control: "control", Extend: "extend", ZWJ: "zwj", Regional_Indicator: "ri", Prepend: "prepend", SpacingMark: "spacingmark", L: "l", V: "v", T: "t", LV: "lv", LVT: "lvt" };
const gb: Range[] = [];
for (const [low, high, property] of parseRanges(process.argv[2] || "/tmp/gbp.txt")) if (classes[property]) gb.push([low, high, classes[property]]);
gb.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
const pictographic: Range[] = parseRanges(process.argv[3] || "/tmp/emoji.txt", "Extended_Pictographic").sort((a, b) => a[0] - b[0] || a[1] - b[1]);
const incb: Range[] = [];
const incbKinds: Record<string, string> = { Linker: "linker", Consonant: "consonant", Extend: "incb_extend" };
for (const raw of readText(process.argv[4] || "/tmp/dcp.txt").split("\n")) {
  const line = raw.split("#")[0].trim();
  if (!line || line.indexOf("InCB") < 0) continue;
  const parts = line.split(";").map(value => value.trim());
  if (parts.length < 3 || parts[1] !== "InCB" || !incbKinds[parts[2]]) continue;
  const bounds = parts[0].split("..");
  incb.push([Number.parseInt(bounds[0], 16), Number.parseInt(bounds.length > 1 ? bounds[1] : bounds[0], 16), incbKinds[parts[2]]]);
}
incb.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
const hex = (value: number) => value.toString(16).toUpperCase();
function emit(name: string, rows: Range[], kind: boolean): string {
  const lines = [`pub const ${name} = [_]${kind ? "GB" : "Range"}{`];
  for (const row of rows) lines.push(kind ? `    .{ .lo = 0x${hex(row[0])}, .hi = 0x${hex(row[1])}, .v = .${row[2]} },` : `    .{ .lo = 0x${hex(row[0])}, .hi = 0x${hex(row[1])} },`);
  lines.push("};");
  return lines.join("\n");
}
const out = [
  "//! GENERATED from UCD GraphemeBreakProperty.txt, emoji-data.txt, and",
  "//! DerivedCoreProperties.txt (InCB) — tools/gen_grapheme.ts. Sorted ranges",
  "//! for the UAX #29 extended grapheme cluster algorithm.", "",
  "pub const Class = enum { other, cr, lf, control, extend, zwj, ri, prepend, spacingmark, l, v, t, lv, lvt };",
  "pub const Incb = enum { linker, consonant, incb_extend };",
  "pub const GB = struct { lo: u21, hi: u21, v: Class };",
  "pub const Range = struct { lo: u21, hi: u21 };",
  "pub const IncbR = struct { lo: u21, hi: u21, v: Incb };", "", emit("gb_class", gb, true), "", emit("ext_pictographic", pictographic, false), "", "pub const incb = [_]IncbR{",
];
for (const row of incb) out.push(`    .{ .lo = 0x${hex(row[0])}, .hi = 0x${hex(row[1])}, .v = .${row[2]} },`);
out.push("};", "");
process.stdout.write(out.join("\n"));
process.stderr.write(`gb=${gb.length} extpict=${pictographic.length} incb=${incb.length}\n`);
